src/storage/secondary/block/dict_block_builder.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use bytes::BufMut;
+use risinglight_proto::rowset::BlockStatistics;
+
+use super::PlainPrimitiveBlockBuilder;
+use crate::array::{Array, I32Array};
+use crate::storage::secondary::block::{BlockBuilder, RleBlockBuilder};
+
+pub(crate) const DICT_NULL_VALUE_KEY: i32 = i32::MIN;
+/// Encodes fixed-width data into a block with dict encoding. The layout is
+/// ```plain
+/// | rle_block_length(u64) |  dict_count_sum (u32) | rle_block | dict_block |
+/// ```
+pub struct DictBlockBuilder<A, B>
+where
+    A: Array,
+    B: BlockBuilder<A>,
+    A::Item: PartialEq,
+    <A::Item as ToOwned>::Owned: Eq + Hash,
+{
+    dict_map: HashMap<<A::Item as ToOwned>::Owned, i32>,
+    data_builder: B,
+    rle_builder: RleBlockBuilder<I32Array, PlainPrimitiveBlockBuilder<i32>>,
+    cur_index: i32,
+}
+
+impl<A, B> DictBlockBuilder<A, B>
+where
+    A: Array,
+    B: BlockBuilder<A>,
+    A::Item: PartialEq,
+    <A::Item as ToOwned>::Owned: Eq + Hash,
+{
+    pub fn new(block_builder: B) -> Self {
+        let builder = PlainPrimitiveBlockBuilder::new(block_builder.estimated_size());
+        let rle_builder =
+            RleBlockBuilder::<I32Array, PlainPrimitiveBlockBuilder<i32>>::new(builder);
+        Self {
+            dict_map: HashMap::new(),
+            data_builder: block_builder,
+            rle_builder,
+            cur_index: DICT_NULL_VALUE_KEY,
+        }
+    }
+}
+
+impl<A, B> BlockBuilder<A> for DictBlockBuilder<A, B>
+where
+    A: Array,
+    B: BlockBuilder<A>,
+    A::Item: PartialEq + Eq + Hash,
+    <A::Item as ToOwned>::Owned: Eq + Hash,
+{
+    fn append(&mut self, item: Option<&A::Item>) {
+        let mut key = DICT_NULL_VALUE_KEY;
+        if let Some(item) = item {
+            if let Some(value) = self.dict_map.get(item) {
+                key = value.to_owned();
+            } else {
+                self.cur_index += 1;
+                key = self.cur_index;
+                self.data_builder.append(Some(item));
+            }
+            self.dict_map.insert(item.to_owned(), key);
```

Comment:
I think we only need to insert in the `else` branch? Otherwise we will need to call `to_owned` multiple times.